components/calendar/src/chinese_based.rs

```diff
@@ -232,76 +414,165 @@ impl<C: ChineseBased + CalendarArithmetic> ChineseBasedDateInner<C> {
-            // This doesn't need to be checked for `None`, since `get_leap_month_from_new_year`
-            // will always return a number greater than or equal to 1, and less than 14.
-            NonZeroU8::new(Self::get_leap_month_from_new_year(first_day_of_year))
-        } else {
-            None
-        };
-        let cache = ChineseBasedCache {
-            new_year: first_day_of_year,
-            next_new_year,
-            leap_month,
-        };
+        let day_before_month_start = data.last_day_of_previous_month(month);
+        let day_of_month = day_of_year - day_before_month_start;
+        let day_of_month = u8::try_from(day_of_month);
+        debug_assert!(day_of_month.is_ok(), "Month too big!");
+        let day_of_month = day_of_month.unwrap_or(1);
+
         // This can use `new_unchecked` because this function is only ever called from functions which
         // generate the year, month, and day; therefore, there should never be a situation where
         // creating this ArithmeticDate would fail, since the same algorithms used to generate the ymd
         // are also used to check for valid ymd.
-        ChineseBasedDateInner(ArithmeticDate::new_unchecked(year, month, day), cache)
+        ChineseBasedDateInner(
+            ArithmeticDate::new_unchecked(extended_year, month, day_of_month),
+            ChineseBasedYearInfo::Data(data),
+        )
+    }
+
+    /// Get a ChineseBasedDateInner from a fixed date, with the related ISO year
+    ///
+    /// Months are calculated by iterating through the dates of new moons until finding the last month which
+    /// does not exceed the given fixed date. The day of month is calculated by subtracting the fixed date
+    /// from the fixed date of the beginning of the month.
+    ///
+    /// The calculation for `elapsed_years` and `month` in this function are based on code from _Calendrical Calculations_ by Reingold & Dershowitz.
+    /// Lisp reference code: https://github.com/EdReingold/calendar-code2/blob/main/calendar.l#L5414-L5459
+    pub(crate) fn chinese_based_date_from_fixed(
+        date: RataDie,
+        iso_year: i32,
+    ) -> ChineseBasedDateInner<C> {
+        // Get the 1-indexed Chinese extended year, used for fetching data from the cache
+        let epoch_as_iso = Iso::iso_from_fixed(C::EPOCH);
+        let mut getter_year = iso_year - epoch_as_iso.year().number + 1;
+
+        let data_option = Self::get_compiled_data_for_year_helper(date, &mut getter_year);
+
+        if let Some(data) = data_option {
+            // cache fetch successful, getter year is just the regular extended year
+            Self::chinese_based_date_from_cached(date, data, getter_year)
+        } else {
+            let (first_day_of_year, next_solstice) =
+                Self::new_year_on_or_before_fixed_date(date, None);
+            let next_new_year = Self::new_year_on_or_before_fixed_date(
+                first_day_of_year + 400,
```

Comment:
nit: 400 or 370? please be consistent

Reply:
I'm going to document the choices

Reply:
In  the process I decided to consolidate all the code dealing with year bounds, which simplifies things greatly